File: c/src/peer_id.c
// SPDX-License-Identifier: MIT
//
// PeerId derivation — C implementation.
//
// Mirrors src/AetherNet.Core/Identity/PeerId.cs and go/identity/peerid.go
// byte-for-byte. Self-contained: depends only on the C standard library and its
// own header (no other repo headers, no POSIX), so the primitive can be lifted
// out and verified standalone.

#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <string.h>

#include "aethernet/peer_id.h"

// Bitcoin base58 alphabet (no 0, O, I, l).
static const char BASE58_ALPHABET[] =
    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

// identity-multihash(code 0x00, len 0x24=36) || protobuf PublicKey(type Ed25519:
// 0x08 0x01; data len 32: 0x12 0x20).
static const uint8_t ED25519_PREFIX[] = {0x00, 0x24, 0x08, 0x01, 0x12, 0x20};
#define ED25519_PREFIX_LEN 6

// 6-byte prefix + 32-byte key.
#define MULTIHASH_LEN (ED25519_PREFIX_LEN + AETHERNET_ED25519_PUBLIC_KEY_LENGTH)

// Divides the big-endian base-256 number in number[firstDigit..len) by 58, in
// place, and returns the remainder. Bytes are treated as unsigned.
static int divmod58(uint8_t *number, size_t first_digit, size_t len) {
    int remainder = 0;
    for (size_t i = first_digit; i < len; i++) {
        int temp = remainder * 256 + (int)number[i];
        number[i] = (uint8_t)(temp / 58);
        remainder = temp % 58;
    }
    return remainder;
}

// Standard base58 (bitcoinj algorithm) — preserves leading zero bytes as leading
// '1's. Writes a NUL-terminated string into out (caller guarantees capacity).
static void base58_encode(const uint8_t *input, size_t input_len, char *out) {
    if (input_len == 0) {
        out[0] = '\0';
        return;
    }

    // Count leading zero bytes.
    size_t zeros = 0;
    while (zeros < input_len && input[zeros] == 0) zeros++;

    // divmod mutates in place, so work on a copy.
    uint8_t buffer[MULTIHASH_LEN];
    memcpy(buffer, input, input_len);

    // Safe upper bound: log(256)/log(58) < 2, so 2 chars per input byte suffices.
    char encoded[MULTIHASH_LEN * 2];
    size_t encoded_len = input_len * 2;
    size_t output_start = encoded_len;

    for (size_t input_start = zeros; input_start < input_len;) {
        encoded[--output_start] =
            BASE58_ALPHABET[divmod58(buffer, input_start, input_len)];
        if (buffer[input_start] == 0) input_start++;  // a digit fully consumed
    }
    // Drop any extra leading '1's the loop produced.
    while (output_start < encoded_len && encoded[output_start] == BASE58_ALPHABET[0]) {
        output_start++;
    }
    // Re-add one '1' per leading zero byte of the input.
    for (; zeros > 0; zeros--) encoded[--output_start] = BASE58_ALPHABET[0];

    size_t out_len = encoded_len - output_start;
    memcpy(out, encoded + output_start, out_len);
    out[out_len] = '\0';
}

bool aethernet_peer_id_from_ed25519(const uint8_t pubkey[32], char out[64]) {
    if (pubkey == NULL || out == NULL) return false;

    uint8_t multihash[MULTIHASH_LEN];
    memcpy(multihash, ED25519_PREFIX, ED25519_PREFIX_LEN);
    memcpy(multihash + ED25519_PREFIX_LEN, pubkey,
           AETHERNET_ED25519_PUBLIC_KEY_LENGTH);

    base58_encode(multihash, MULTIHASH_LEN, out);
    return true;
}

File: src/AetherNet.Core/Identity/PeerId.cs
// SPDX-License-Identifier: MIT

using System;

namespace AetherNet.Identity;

/// <summary>
/// Derives a libp2p <b>PeerID</b> from a node's Ed25519 public key — the bridge between an
/// AetherNet identity and the global libp2p relay / DHT used by the decentralised relay layer.
///
/// <para>Because AetherNet and libp2p both key identity off the same Ed25519 public key, the PeerID
/// is a <em>pure, deterministic</em> function of that key — no lookup table, no network. A node can
/// compute its own PeerID (to announce on the libp2p DHT) and any peer's PeerID (to dial it) from
/// the public key alone.</para>
///
/// <h3>Encoding (must be byte-identical across every SDK language)</h3>
/// <list type="number">
///   <item><description><b>protobuf PublicKey</b> = <c>08 01</c> (field 1 Type = Ed25519) <c>12 20</c>
///     (field 2 Data, length 32) followed by the 32-byte key — 36 bytes total.</description></item>
///   <item><description><b>identity multihash</b> = <c>00</c> (identity hash code) <c>24</c> (length 36)
///     followed by the protobuf — 38 bytes. libp2p uses the identity multihash for keys whose
///     serialized form is ≤ 42 bytes, which Ed25519 always is.</description></item>
///   <item><description><b>PeerID string</b> = base58btc (Bitcoin alphabet) of the 38-byte multihash,
///     WITHOUT a multibase prefix. Always renders as <c>12D3Koo…</c> for Ed25519.</description></item>
/// </list>
///
/// Verified byte-for-byte against real <c>js-libp2p</c> output; see <c>fixtures/peerid/</c>.
/// </summary>
public static class PeerId
{
    // Bitcoin base58 alphabet (no 0, O, I, l).
    private const string Base58Alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

    // identity-multihash(code 0x00, len 0x24=36) || protobuf PublicKey(type Ed25519: 0x08 0x01; data len 32: 0x12 0x20)
    private static readonly byte[] Ed25519Prefix = { 0x00, 0x24, 0x08, 0x01, 0x12, 0x20 };

    /// <summary>Length in bytes of a raw Ed25519 public key.</summary>
    public const int Ed25519PublicKeyLength = 32;

    /// <summary>
    /// Returns the libp2p PeerID string (e.g. <c>12D3Koo…</c>) for a 32-byte Ed25519 public key.
    /// </summary>
    /// <exception cref="ArgumentNullException"><paramref name="publicKey"/> is null.</exception>
    /// <exception cref="ArgumentException"><paramref name="publicKey"/> is not exactly 32 bytes.</exception>
    public static string FromEd25519PublicKey(byte[] publicKey)
    {
        ArgumentNullException.ThrowIfNull(publicKey);
        if (publicKey.Length != Ed25519PublicKeyLength)
            throw new ArgumentException($"Ed25519 public key must be {Ed25519PublicKeyLength} bytes, got {publicKey.Length}.", nameof(publicKey));

        var multihash = new byte[Ed25519Prefix.Length + Ed25519PublicKeyLength];
        Buffer.BlockCopy(Ed25519Prefix, 0, multihash, 0, Ed25519Prefix.Length);
        Buffer.BlockCopy(publicKey, 0, multihash, Ed25519Prefix.Length, Ed25519PublicKeyLength);
        return Base58Encode(multihash);
    }

    // Standard base58 (bitcoinj algorithm) — preserves leading zero bytes as leading '1's.
    private static string Base58Encode(byte[] input)
    {
        if (input.Length == 0) return string.Empty;

        int zeros = 0;
        while (zeros < input.Length && input[zeros] == 0) zeros++;

        var buffer = (byte[])input.Clone(); // divmod mutates in place
        var encoded = new char[input.Length * 2]; // safe upper bound
        int outputStart = encoded.Length;

        for (int inputStart = zeros; inputStart < buffer.Length;)
        {
            encoded[--outputStart] = Base58Alphabet[DivMod(buffer, inputStart, 256, 58)];
            if (buffer[inputStart] == 0) inputStart++; // a digit fully consumed
        }
        // Drop extra leading '1's the loop may have produced.
        while (outputStart < encoded.Length && encoded[outputStart] == Base58Alphabet[0]) outputStart++;
        // Re-add one '1' per leading zero byte of the input.
        for (; zeros > 0; zeros--) encoded[--outputStart] = Base58Alphabet[0];

        return new string(encoded, outputStart, encoded.Length - outputStart);
    }

    // Divides the big-endian base-256 number in number[firstDigit..] by 58, in place, returns the remainder.
    private static int DivMod(byte[] number, int firstDigit, int baseIn, int baseOut)
    {
        int remainder = 0;
        for (int i = firstDigit; i < number.Length; i++)
        {
            int digit = number[i] & 0xFF;
            int temp = remainder * baseIn + digit;
            number[i] = (byte)(temp / baseOut);
            remainder = temp % baseOut;
        }
        return remainder;
    }
}
